contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -3435,3 +3510,415 @@ objectproperty_internal(PG_FUNCTION_ARGS)
+	if (tupdesc->natts < 2)
+		ereport(ERROR,
+				(errcode(ERRCODE_SYNTAX_ERROR),
+				 errmsg("query-specified return tuple and " \
+						"bbf_pivot function are not compatible")));
+
+	/* load up the categories hash table */
+	bbf_pivot_hash = load_categories_hash(bbf_pivot_cat_sql, per_query_ctx);
+
+	/* let the caller know we're sending back a tuplestore */
+	rsinfo->returnMode = SFRM_Materialize;
+
+	/* now go build it */
+	rsinfo->setResult = get_bbf_pivot_tuplestore(bbf_pivot_src_sql,
+												bbf_pivot_hash,
+												tupdesc,
+												rsinfo->allowedModes & SFRM_Materialize_Random);
+
+	/*
+	 * SFRM_Materialize mode expects us to return a NULL Datum. The actual
+	 * tuples are in our tuplestore and passed back through rsinfo->setResult.
+	 * rsinfo->setDesc is set to the tuple description that we actually used
+	 * to build our tuples with, so the caller can verify we did what it was
+	 * expecting.
+	 */
+	rsinfo->setDesc = tupdesc;
+	MemoryContextSwitchTo(oldcontext);
+
+	oldcontext = MemoryContextSwitchTo(tsql_outmost_context);
+	/* TODO: do we need to free the list */
+	tsql_outmost_estat->pivot_parsetree_list = list_delete_nth_cell(tsql_outmost_estat->pivot_parsetree_list, tsql_outmost_estat->pivot_number - 1);
+	tsql_outmost_estat->pivot_number--;
+	MemoryContextSwitchTo(oldcontext);
+	return (Datum) 0;
+}
+
+/*
+ * load up the categories hash table
+ */
+static HTAB *
+load_categories_hash(RawStmt *cats_sql, MemoryContext per_query_ctx)
+{
+	HTAB	   *bbf_pivot_hash;
+	HASHCTL		ctl;
+	int			ret;
+	uint64		proc;
+	MemoryContext SPIcontext;
+
+	/* initialize the category hash table */
+	ctl.keysize = MAX_CATNAME_LEN;
+	ctl.entrysize = sizeof(bbf_pivot_HashEnt);
+	ctl.hcxt = per_query_ctx;
+
+	/*
+	 * use INIT_CATS, defined above as a guess of how many hash table entries
+	 * to create, initially
+	 */
+	bbf_pivot_hash = hash_create("bbf_pivot hash",
+								INIT_CATS,
+								&ctl,
+								HASH_ELEM | HASH_STRINGS | HASH_CONTEXT);
+
+	/* Connect to SPI manager */
+	if ((ret = SPI_connect()) < 0)
+		/* internal error */
+		elog(ERROR, "load_categories_hash: SPI_connect returned %d", ret);
+
+	/* Retrieve the category name rows */
+	ret = SPI_execute_raw_parsetree(cats_sql, true, 0);
+	proc = SPI_processed;
+
+	/* Check for qualifying tuples */
+	if ((ret == SPI_OK_SELECT) && (proc > 0))
+	{
+		SPITupleTable *spi_tuptable = SPI_tuptable;
+		TupleDesc	spi_tupdesc = spi_tuptable->tupdesc;
+		uint64		i;
+
+		/*
+		 * The provided categories SQL query must always return one column:
+		 * category - the label or identifier for each column
+		 */
+		if (spi_tupdesc->natts != 1)
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					 errmsg("provided \"categories\" SQL must " \
+							"return 1 column of at least one row")));
+
+		for (i = 0; i < proc; i++)
+		{
+			bbf_pivot_cat_desc *catdesc;
+			char	   *catname;
+			HeapTuple	spi_tuple;
+
+			/* get the next sql result tuple */
+			spi_tuple = spi_tuptable->vals[i];
```

Comment:
You are looping from `i=0` to `i<proc` and accessing `spi_tuptable->vals[i]`.
Make sure you dont go out of bounds. I suggest add Assertion here.

```
Assert(spi_tuptable->numvals >= proc);
```

Reply:
SPI_tuptable->values,  SPI_tuptable->numvals, and SPI_processed (proc) are managed by the same SPI_Current context, and SPI_processed is always equals to SPI_tuptable->numvals. So there is no need to do extra assertion.